Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/safetynet/SafetyNetException.kt

```diff
@@ -1,6 +1,32 @@
 package de.rki.coronawarnapp.datadonation.safetynet
 
 class SafetyNetException constructor(
-    message: String?,
+    val type: Type,
+    message: String,
     cause: Throwable? = null
-) : Exception(message, cause)
+) : Exception("$type: $message", cause) {
+
+    enum class Type {
```

Comment:
would it make sense, to make the enum available for the whole data donation, e.g. in an own class? ->could be reused for the survey e.g. (not only in the SafetyNetException)

Reply:
> would it make sense, to make the enum available for the whole data donation, e.g. in an own class?

This is available for the whole data donation, isn't it?
Both survey and analytics use safetynet, and if they receive an exception they can check the error type if they want to perform specific behavior depending on what error it is.
As this is an error type specific to SafetyNet and only available via `SafetyNetException`s being thrown it makes sense to keep it here imho.

> could be reused for the survey e.g. (not only in the SafetyNetException)

According to the tech spec, these error codes are specifically related to safetynet behavior. Did I missunderstand the spec or are you suggestion to put all error types (not just safetynet) into the same enum?

Reply:
"_are you suggestion to put all error types (not just safetynet) into the same enum?_"
yes, that was the idea. Just one allover enum class for all error codes regarding dataDonation...

Reply:
Hm, I'm not sure when this would be an advantage. Which class would have to handle all types of errors? By scoping it we have more robust error handling, e.g. SafetyNet errors will only popup in specific locations in survey and analytics core logic, as the enum error types are specific to SafetyNet and need to be handled in the survey and analytics class, and not further downstream.

The current typing would allow `ImaginarySurveyRepo` to just use a `when(error.type) { }` and let the compiler enforce that all error types that can happen during a SafetyNet check are handled as desired.

Any additional exceptions e.g. `SurveyLogicSaysNoException` that occur in the survey related logic could have their own exception typing if necessary, and could also include the `SafetyNetException` as cause, but I'd expect any classes downstream of `ImaginarySurveyRepo` to not need to be aware of the ways that SafetyNet can fail. This should be a hidden implementation detail of the class that uses the SafetyNet code.

:thinking:  a bit like "per feature not per layer", i.e. group errors by feature/module, not by layer, i.e. "all errors in one package".

Reply:
I think, both ways are possible.
It won't stop the train..